pep-9999.rst

```diff
@@ -0,0 +1,891 @@
+Linux and BSD* distribution packagers. This has triggered several license-related
+discussions and issues and in particular:
+
+- https://github.com/pypa/warehouse/issues/2996
+- https://github.com/pypa/interoperability-peps/issues/46
+- https://github.com/pypa/packaging-problems/issues/41
+- https://github.com/pypa/wheel/issues/138
+- https://github.com/pombredanne/spdx-pypi-pep/issues/1
+
+On average, Python packages tend to have more ambiguous or missing license
+information than other common application package formats (such as npm, Maven or
+Gem) as can be seen in the statistics [#cdstats]_ page of the ClearlyDefined
+[#cd]_ project that cover all packages from PyPI, Maven, npm and Rubygems.
+ClearlyDefined is an open source project to help improve clarity of other open
+source projecs that is incubating at the OSI (Open Source Initiative) [#osi]_.
+
+
+Rationale
+=========
+
+A mini survey of existing license metadata definitions in use in Python today
+and documented in several other system/distro and application package formats is
+provided in Appendix 2, of this PEP.
+
+There are a few takeaways from the survey:
+
+- Most package formats use a single `License` field.
+
+- Many modern package formats use some form of license expression syntax to
+  optionally combine more than one license identifiers together. SPDX and
+  SPDX-like syntaxes are the most popular in use.
+
+- SPDX license IDs are becoming a de-facto way to reference common licenses
+  everywhere, whether or not a license expression syntax is used.
+
+- Several package formats support documenting both a license expression and
+  the paths of the corresponding files that contain the license text. Most free
+  and open source software licenses require to include their full text in a
+  distribution.
+
+These considerations have guided the design and recommendations of this PEP.
+
+The reuse of the `License` field with license expressions will provide an
+intuitive and more structured way to express the license of a distribution using
+a well-defined syntax and well-known license ID.
+
+Over time, recommending the usage of these expressions will help Python package
+publishers improve the clarity of their license documentation to the benefit of
+packages authors, consumers and redistributors.
+
+
+Core Metadata Specification updates
+===================================
+
+The canonical source for the names and semantics of each of the supported
+metadata fields is the Core Metadata Specification [#cms]_ document.
+
+The details of the updates considered to the Core Metadata Specification [#cms]_
+document as part of this PEP are detailed here and will be added to the
+canonical source once this PEP is approved.
+
+
+Added in Version 2.2
+--------------------
+
+License-File (multiple use)
+:::::::::::::::::::::::::::
+
+The License-File is a string that is a package-root relative path to a license
+file. The license file content __must__ be UTF-8-encoded text.
```

Comment:
Counter-proposal: it is a simple filename for a license file in the dist-info directory.

I see license as part of the metadata, not the program code, so it seems it would be better to have all the info inside dist-info.

(Your program wants to access the license file to display it?  use functions in `importlib.metadata`)

Reply:
On the mailing list, it was suggest to reuse the rules for RECORD, so with that it would be a path relative to site-packages directory, which makes it easy to point to `project-0.42.dist-info/license.txt`.

Reply:
@merwork this also highlights another issue. With this setup.cfg
```
[metadata]
license_files= 
 LICENSE
 license
```
A whee will have both:
```
 LICENSE
 license
```
That's on a POSIX filesystem that's case-sensitive for paths.
But if that wheel were installed on a case-insensitive FS such as Windows and more recently macOS APFS, it may be that one of te two files get overwritten which is IMHO not an acceptable solution for METADATA. IMHO we need to specify that each `License-File` entry must be unique ignoring case. And that will be for tools to honor (and for `wheel` to be fixed accordingly) 

@agronholm and @pfmoore  what's your take on this?

Reply:
Note that "the paths of license files in the source tree" is a different information than "the path of the license files copied to the dist info directory". Hence while setup.cfg has  license_files, it is not the same information as will be stored in the dist-info metadata.

Reply:
IMO what gets put into the wheel is not a matter for this PEP - I'd see it as a bug if the `[metadata]` section in `setup.cfg` can contain two different files, and not all of those files end up in the wheel (for whatever reason - stripping of directory names, case sensitivity issues, whatever). There may be a wheel spec issue around whether wheels are allowed to contain two files whose names differ only in case - that's a question for the wheel spec if so (my assumption would be that it shouldn't be allowed).

But for this PEP, all that matters is the resulting metadata, as @hroncok states. The huge problem here is that currently the [core metadata spec](https://packaging.python.org/specifications/core-metadata/) doesn't have any fields that refer to external files, and doesn't even include a concept of "other files" by which such metadata could be defined. Consider for example the proposed extension to the simple repository API, to expose package metadata as a separate object. This would mean that tools could access the `License-Files` metadata field, but not have any of the other content of the distribution available.

I didn't spot this when `License-File` was added to the PEP, sorry. To be honest, I'd strongly recommend *not* adding this field, as it opens a whole can of worms about how do we handle the fact that the metadata now needs additional files to be usable. But if you want to do so, I think we need to take that question back to Discourse and thrash it out as a general question, not related to this PEP, "How should package metadata reference additional files?"

The discussions about adding metadata to `pyproject.toml` touched on this, because we noted that the *only* metadata object that had long-form data was `Description`, and that's handled via a special exception (it's located as the document content, not as a header - see [here](https://packaging.python.org/specifications/core-metadata/#description) - "Alternatively, the distribution’s description may instead be provided in the message body"). So that PEP will be potentially impacted by this question.

Reply:
There is one dist-info file that’s not defined by PEP 376: entry_points.txt
https://packaging.python.org/specifications/entry-points/#file-format

It is not referenced from METADATA but is present in RECORD, so maybe it’s precedent enough!
(the naming doesn’t use the all-caps style because the spec was retrofitted from a setuptools invention)

Reply:
Good point, it works as a precedent for shipping license files - but as you say, it doesn't indicate how we reference other files from metadata. Also (and this is something else the proposal here needs to consider) what if someone tries to ship a license file called `entry-points.txt`? Yes, I know it's a silly thing to do, but standards need to cover edge cases...

I think if we're going to standardise shipping license *files* (as opposed to the original scope of this PEP which was just about specifying what license was in use), we probably need to reserve a namespace in `.dist-info` - say that all licenses must go under `.dist-info/licenses`, or something.

On a procedural note, by the way, this discussion is getting too complex to be handled just on the tracker, it should be part of the main discussion thread. @pombredanne could you summarise the discussion so far, and post that summary to the Discourse thread to allow others to comment? Thanks.

Reply:
@pfmoore Let me summarize all the discussion over the week-end and start a new discourse topic.

> what if someone tries to ship a license file called entry-points.txt

FWIW, you can do that alright but it is even worse than that: I can overwrite the `dist-info/METADATA` with a `METADATA` arbitrary file entry in the `license_files` setup.cfg section  :| 
This is a wheel bug alright, and I will tackle this as that and that's not a topic for this PEP.

What's a topic is that `License-File` as it is specified is NOT a solution for sure.

Reply:
> Counter-proposal: it is a simple filename for a license file in the dist-info directory.

I actually think this is a much better approach than including the license file.

A case that comes to mind is a Sphinx theme I'm working on which is an MIT license, and vendors material-icons with it's own different license in the same directory. Referencing the specific files in the distribution is much better IMO, since I would like my metadata to not indicate that the entire project is copyrighted by Google. :)

I think what's in the PEP right now is a much more capable mechanism to handle such instances.

Reply:
It's not that simple. Some licences are applied by using a comment header and attaching multiple files (e.g. LGPL needs a file with full GPL text and a file with full LGPL text, some other licenses need LICENSE and NOTICE). While merging the text to one file is technically possible, it's non-standard.